### PR TITLE
Simplex: DataOut::attach_triangulation()

### PR DESCRIPTION
**Pull request by peterrum**
depends on #11924

### CODE REVIEW

File: include/deal.II/numerics/data_out_dof_data.templates.h [hunk 1]
```diff
@@ -2213,10 +2214,39 @@ DataOut_DoFData<DoFHandlerType, patch_dim, patch_space_dim>::get_fes() const
...
+
+      // TODO: below we select linear elements for simplex, wedge, and pyramid;
+      // we would like to select constant functions, but that is not implemented
+      // yet
+      for (const auto reference_cell : reference_cells)
+        {
+          if ((reference_cell == dealii::ReferenceCells::Vertex) ||
+              (reference_cell == dealii::ReferenceCells::Line) ||
+              (reference_cell == dealii::ReferenceCells::Quadrilateral) ||
+              (reference_cell == dealii::ReferenceCells::Hexahedron))
```
Comment (bangerth) [inline include/deal.II/numerics/data_out_dof_data.templates.h:2230]: `if (reference_cell.is_hyper_cube())` :-)

File: include/deal.II/numerics/data_out_dof_data.templates.h [hunk 2]
```diff
@@ -2213,10 +2214,39 @@ DataOut_DoFData<DoFHandlerType, patch_dim, patch_space_dim>::get_fes() const
...
-      finite_elements.resize(1);
-      finite_elements[0] =
-        std::make_shared<dealii::hp::FECollection<dhdim, dhspacedim>>(
-          FE_DGQ<dhdim, dhspacedim>(0));
+      Assert(triangulation != nullptr, ExcNotImplemented());
+
+      const auto reference_cells = triangulation->get_reference_cells();
+      finite_elements.reserve(reference_cells.size());
+
+      // TODO: below we select linear elements for simplex, wedge, and pyramid;
```
Comment (bangerth) [inline include/deal.II/numerics/data_out_dof_data.templates.h:2222]:
```suggestion
      // TODO: below we select linear, discontinuous elements for simplex, wedge, and pyramid;
```

File: include/deal.II/numerics/data_out_dof_data.templates.h [hunk 3]
```diff
@@ -2213,10 +2214,39 @@ DataOut_DoFData<DoFHandlerType, patch_dim, patch_space_dim>::get_fes() const
...
+        {
+          if ((reference_cell == dealii::ReferenceCells::Vertex) ||
+              (reference_cell == dealii::ReferenceCells::Line) ||
+              (reference_cell == dealii::ReferenceCells::Quadrilateral) ||
+              (reference_cell == dealii::ReferenceCells::Hexahedron))
+            finite_elements.emplace_back(
+              std::make_shared<dealii::hp::FECollection<dhdim, dhspacedim>>(
+                FE_DGQ<dhdim, dhspacedim>(0)));
+          else if ((reference_cell == dealii::ReferenceCells::Triangle) ||
+                   (reference_cell == dealii::ReferenceCells::Tetrahedron))
```
Comment (bangerth) [inline include/deal.II/numerics/data_out_dof_data.templates.h:2235]: is_simplex

File: tests/simplex/data_out_write_vtk_03.cc [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  DataOut<dim> data_out;
+
+  data_out.attach_triangulation(tria);
+
+  data_out.build_patches();
+
+#if false
+  std::ofstream output("test." + std::to_string(dim) + ".vtk");
+  data_out.write_vtk(output);
+#else
```
Comment (bangerth) [inline tests/simplex/data_out_write_vtk_03.cc:60]: Just remove this part

**Review by bangerth — COMMENTED** [review]
OK after you merge the other PR.

File: tests/simplex/data_out_write_vtk_03.cc (hunk 4 above)
Comment (peterrum) [reply tests/simplex/data_out_write_vtk_03.cc:60]: I have made the changes. This one I keep. This something we are doing quite commonly in the tests. It gives us the opportunity to quickly check the output in Paraview if we come back to the test.

**Review by bangerth — APPROVED** [review]

**Pull request merged**